File: examples/scalar_udf/java/app/src/main/java/scalar_udf/CelsiusToFahrenheit.java
/**
 * Copyright (c) 2026 Jeffrey Jonathan Jennings
 *
 * @author Jeffrey Jonathan Jennings (J3)
 *
 *
 */
package scalar_udf;

import org.apache.flink.table.functions.ScalarFunction;

/**
 * A Flink scalar user-defined function (UDF) that converts a temperature value
 * from degrees Celsius to degrees Fahrenheit.
 *
 * <p>Being a scalar UDF, it is invoked once per input row and returns exactly
 * one value per invocation. The function is stateless and deterministic: the
 * same input will always produce the same output, which allows Flink's planner
 * to safely cache, reorder, and replay calls during failure recovery.
 *
 * <h3>SQL invocation</h3>
 * <pre>{@code
 * CREATE FUNCTION CelsiusToFahrenheit
 *     AS 'scalar_udf.CelsiusToFahrenheit'
 *     LANGUAGE JAVA;
 *
 * SELECT temperature_in_celsius,
 *        CelsiusToFahrenheit(temperature_in_celsius) AS temperature_in_fahrenheit
 *   FROM readings;
 * }</pre>
 */
public class CelsiusToFahrenheit extends ScalarFunction {

    /**
     * Converts a Celsius temperature to its Fahrenheit equivalent using the
     * formula {@code F = (C * 9/5) + 32}.
     *
     * <p>Flink discovers scalar UDF logic by reflecting on public {@code eval}
     * methods, so the method name and visibility must remain as declared.
     *
     * @param celsius the temperature in degrees Celsius; may be {@code null}
     *                to represent a SQL {@code NULL} value
     * @return the temperature in degrees Fahrenheit, or {@code null} if the
     *         input was {@code null} (preserving SQL {@code NULL} semantics)
     */
    public Double eval(Double celsius) {
        if (celsius == null)
            return null;
        return (celsius * 9.0 / 5.0) + 32.0;
    }
}

File: examples/scalar_udf/python/celsius_to_fahrenheit.py
"""
Copyright (c) 2026 Jeffrey Jonathan Jennings

Author: Jeffrey Jonathan Jennings (J3)
"""
from typing import Optional

from pyflink.table.udf import ScalarFunction, udf
from pyflink.table import DataTypes


class CelsiusToFahrenheit(ScalarFunction):
    """
    A Flink scalar user-defined function (UDF) that converts a temperature value
    from degrees Celsius to degrees Fahrenheit.

    Being a scalar UDF, it is invoked once per input row and returns exactly one
    value per invocation. The function is stateless and deterministic: the same
    input always produces the same output, which allows Flink's planner to
    safely cache, reorder, and replay calls during failure recovery.

    SQL invocation::

        t_env.create_temporary_system_function(
            "CelsiusToFahrenheit",
            celsius_to_fahrenheit,
        )

        t_env.execute_sql('''
            SELECT temperature_in_celsius,
                   CelsiusToFahrenheit(temperature_in_celsius) AS temperature_in_fahrenheit
              FROM readings
        ''').print()
    """

    def eval(self, celsius: Optional[float]) -> Optional[float]:
        """
        Convert a Celsius temperature to its Fahrenheit equivalent using the
        formula ``F = (C * 9/5) + 32``.

        PyFlink discovers scalar UDF logic by looking for an ``eval`` method on
        the ``ScalarFunction`` subclass, so the method name must remain as
        declared.

        :param celsius: temperature in degrees Celsius; may be ``None`` to
            represent a SQL ``NULL`` value.
        :return: temperature in degrees Fahrenheit, or ``None`` if the input
            was ``None`` (preserving SQL ``NULL`` semantics).
        """
        if celsius is None:
            return None
        return (celsius * 9.0 / 5.0) + 32.0


celsius_to_fahrenheit = udf(
    CelsiusToFahrenheit(),
    input_types=[DataTypes.DOUBLE()],
    result_type=DataTypes.DOUBLE(),
)
